### PR TITLE
More detailed errors for invalid array lengths (such as division by zero)

### LINKED ISSUE #1259: an inconsistent error message for array size expressions
At version `0.4.3-develop.2016.10.20+commit.9d304501.Linux.g++`, an array size can be a sum of constants, but cannot be a division by zero.  In case of division by zero, the error message says that the compiler expected an integer literal.

Either the index expression should be limited to an integer literal, or the error message should be changed to "expected a constant expression whose value is a non-negative integer."

```
contract T {
    function f() {
        int[3+4] memory a; //succeeds
    }
    function g() {
        int[3/0] memory a; // fails
    }
}
```

```
test.sol:6:13: Error: Invalid array length, expected integer literal.
        int[3/0] memory a;
            ^-^
```

### PR DESCRIPTION
**Pull request by axic**
Fixes #1259.

### CODE REVIEW

File: libsolidity/analysis/ConstantEvaluator.cpp [hunk 1]
```diff
@@ -46,9 +48,24 @@ void ConstantEvaluator::endVisit(BinaryOperation const& _operation)
 	if (!dynamic_cast<RationalNumberType const*>(rightType.get()))
 		m_errorReporter.fatalTypeError(_operation.rightExpression().location(), "Invalid constant expression.");
 	TypePointer commonType = leftType->binaryOperatorResult(_operation.getOperator(), rightType);
-	if (Token::isCompareOp(_operation.getOperator()))
```
Comment (chriseth) [inline libsolidity/analysis/ConstantEvaluator.cpp:49]: This is a slight change in semantics. Did you test that against e.g. `bool constant x = 2**700 > 2**600;`?

Comment (axic) [reply libsolidity/analysis/ConstantEvaluator.cpp:49]: Fails, but in the non-constant case too:

```
3107.sol:2:21: Error: Operator > not compatible with types int_const 5260135901548373507240989882880128665550339802823173859498280903068732154297080822113666536277588451226982968856178217713019432250183803863127814770651880849955223671128444598191663757884322717271293251735781376 and int_const 4149515568880992958512407863691161151012446232242436899995657329690652811412908146399707048947103794288197886611300789182395151075411775307886874834113963687061181803401509523685376
  bool constant x = 2**700 > 2**600;
                    ^-------------^
```

Before this change it works for constants but not anything else. I think that should be consistent (ie. working in both cases).

Comment (chriseth) [reply libsolidity/analysis/ConstantEvaluator.cpp:49]: ok!

**Comment by chriseth** [comment]
Conflict in changelog.

**Review by chriseth — APPROVED** [review]

File: libsolidity/analysis/ConstantEvaluator.cpp (hunk 1 above)
Comment (axic) [reply libsolidity/analysis/ConstantEvaluator.cpp:49]: Wait a second, this code is not even used for such constants. It is only used to determine the length of an array.

Comment (chriseth) [reply libsolidity/analysis/ConstantEvaluator.cpp:49]: yes

**Pull request merged**